File: python/Rooky.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#
"""Модуль для работы с манипулятором Rooky."""

__author__ = "Promobot"
__license__ = "Apache License, Version 2.0"
__status__ = "Production"
__url__ = "https://git.promo-bot.ru"
__version__ = "0.1.0"


import bus_handler
import time
import Servo
import Servo_ppm

_right_arm_servos = [21, 22, 23, 24, 25]
_left_arm_servos = [31, 32, 33, 34, 35]
_right_arm_ppm_addr = 26
_left_arm_ppm_addr = 36

joint_ratio = {}
joint_limits = {}

joint_ratio['joint_1'] = 3
joint_ratio['joint_2'] = 1.9
joint_ratio['joint_3'] = 1.9
joint_ratio['joint_4'] = 1.2
joint_ratio['joint_5'] = 1.2
joint_ratio['joint_6'] = 1.7
joint_ratio['joint_7'] = 0.37

joint_limits['joint_1'] = [-25, 134]
joint_limits['joint_2'] = [0, 83]
joint_limits['joint_3'] = [-90, 83]
joint_limits['joint_4'] = [0, 80]
joint_limits['joint_5'] = [-86, 86]
joint_limits['joint_6'] = [-20, 31]
joint_limits['joint_7'] = [0, 74]

_rooky_parts = ['left', 'right']

class Rooky():
	"""Класс для работы с манипулятором.

	Args:
		* port (str): имя посдеовательного порта манипулятора.
		* side (str): Типа манипулятора. Левый или правый.
		* debug (bool): Вывод отладочной информации в консоль.
	
	"""
	def __init__(self, port, side = "left", debug = False):
		self._debug = debug
		self._rooky_side = side
		self._arm_servos = []
		self._arm_ppm_servos = []
		self._servos_id_list = []
		self._cur_setp = {}
		self._cur_setp['joint_2'] = 0
		self._cur_setp['joint_3'] = 0
		self._cur_setp['joint_4'] = 0
		self._cur_setp['joint_5'] = 0

		if side in _rooky_parts:

			self.master = bus_handler.Bus(port = port, baudrate = 460800, debug = debug, timeout = 1.0)

			if side == 'right':
				self._servos_id_list = _right_arm_servos
				self._ppm_addr = _right_arm_ppm_addr
				self._ppm_start_pos = (0, 90)

			if side == 'left':
				self._servos_id_list = _left_arm_servos
				self._ppm_addr = _left_arm_ppm_addr   
				self._ppm_start_pos = (0, -90)


			for i in self._servos_id_list:
				self._arm_servos.append(Servo.Servo(i,self.master.bus))

			self._arm_ppm_servos.append(Servo_ppm.Servo_ppm(self._ppm_addr, self.master.bus, 1, self._ppm_start_pos[0]))
			self._arm_ppm_servos.append(Servo_ppm.Servo_ppm(self._ppm_addr, self.master.bus, 2, self._ppm_start_pos[1]))

			for servo in self._arm_servos:	
				servo.set_command(0x0)
				servo.set_speed(10)
				servo.set_torque(True)
				servo.set_point(0)

			self._untouched_values = self._arm_ppm_servos[0].read_touch()
		time.sleep(3)		


	def _degrees_to_ticks(self,deg):
		return deg / 0.0219


	def _check_limits(self,name,position):
		if joint_limits[name][0] <= position <= joint_limits[name][1]:
			return 1
		else:
			print("ERROR! %s: invalid setpoint: %d" % (name,position)) 

	def _move_simple_joint(self,name,speed,position):
		if (self._check_limits(name,position)):

			setpoint = int(self._degrees_to_ticks(position) * joint_ratio[name])

			self._arm_servos[0].set_torque(True)

			self._arm_servos[0].set_speed(speed)

			self._arm_servos[0].set_point(setpoint)
		else:
			return -1

	def _move_diff_joint(self,name,speed):
		if name == 'joint_2' or name == 'joint_3':
			self._arm_servos[1].set_speed(speed)
			self._arm_servos[2].set_speed(speed)

			self._arm_servos[1].set_torque(True)
			self._arm_servos[2].set_torque(True)

			self._arm_servos[1].set_point(self._cur_setp['joint_2'] + self._cur_setp['joint_3'])
			self._arm_servos[2].set_point(self._cur_setp['joint_2'] - self._cur_setp['joint_3'])
		if name == 'joint_4' or name == 'joint_5':
			self._arm_servos[3].set_speed(speed)
			self._arm_servos[4].set_speed(speed)

			self._arm_servos[3].set_torque(True)
			self._arm_servos[4].set_torque(True)

			self._arm_servos[3].set_point(self._cur_setp['joint_4'] + self._cur_setp['joint_5'])
			self._arm_servos[4].set_point(self._cur_setp['joint_4'] - self._cur_setp['joint_5'])


	def _set_diff_joint(self,name,speed,position):
		if (self._check_limits(name,position)):
			self._cur_setp[name] = int(self._degrees_to_ticks(position) * joint_ratio[name])
			self._move_diff_joint(name,speed)
		else:
			return -1

	def _move_servo_joint(self, name, position):
		if name == 'joint_6':
			self._arm_ppm_servos[0].set_point(position * joint_ratio[name])
		elif name == 'joint_7' and self._rooky_side is "left":
			self._arm_ppm_servos[1].set_point(self._ppm_start_pos[1] + position * joint_ratio[name])
		elif name == 'joint_7' and self._rooky_side is "right":
			self._arm_ppm_servos[1].set_point(self._ppm_start_pos[1] - position * joint_ratio[name])

	def _check_erros(self,data):
		if len(data["Errors"]) > 0:
			print("ERROR! Servo: %d: in troubles: %s" % (data["ID"],data["Errors"])) 

	def move_joint(self,joint_name,speed = 10,position = 0):
		"""Перемещение сустава в желаемую позицию.

		Args:
			* joint_name (str):имя сустава.
			* speed (float): Скорость перемещения сустава.
			* position (float): Желаемая позиция сустава в градусах.
		
		"""
		if joint_name == 'joint_1':
			self._move_simple_joint(joint_name, speed, position)
		if joint_name == 'joint_2':
			self._set_diff_joint(joint_name, speed, position)
		if joint_name == 'joint_3':
			self._set_diff_joint(joint_name, speed, position) 
		if joint_name == 'joint_4':
			self._set_diff_joint(joint_name, speed, position)       
		if joint_name == 'joint_5':
			self._set_diff_joint(joint_name, speed, position)
		if joint_name == 'joint_6':
			self._move_servo_joint(joint_name, position)
		if joint_name == 'joint_7':
			self._move_servo_joint(joint_name, position)

	def read_servos(self):
		"""Чтение доступных данных с сервоприводов
		Args:
			None
		Returns:
			* список словарей с информацией от сервоприводов.
			См Servo.get_data(), а также Servo_ppm.Servo_ppm.get_data()
		Raises:
			None
		"""
		_servo_data = [None] * len(self._arm_servos)
		i = 0
		for servo in self._arm_servos:
			_servo_data[i] = servo.get_data()
			if _servo_data[i]:
				self._check_erros(_servo_data[i])
			else:
				print(self._servos_id_list)
				print("ERROR! Cant read servo: %d" % (self._servos_id_list[i])) 
			i = i + 1
		for ppm in self._arm_ppm_servos:
			_servo_data.append(ppm.get_data())
		return _servo_data

	def read_touch(self):
		"""Чтение данных с датчиков касания

        Args:
            None
        Returns:
            * Словарь с ключами:
                | "Touch_1"
                | "Touch_2"
				или пустой, при ошибке

        Raises:
            None

        """
		return self._arm_ppm_servos[0].read_touch()

	def is_touched(self):
		"""Есть ли факт касания датчика касания

        Args:
            None
        Returns: 
			* True - есть касание датчика
			  False - нет касания датчика

        """
		vals = self._arm_ppm_servos[0].read_touch()
		if vals["Touch_1"] < self._untouched_values["Touch_1"] * 0.95 or vals["Touch_2"] < self._untouched_values["Touch_2"] * 0.95:
			return True
		else:
			return False

	def relax(self,state):
		"""Полностью (включая тормоз) отключить питание двигателей всех суставов.

		Args:
			* state (bool):True - отключить питание. False - включить питание.
		
		"""
		if state:
			for servo in self._arm_servos:
				servo.set_command(0xAAAA)
			self._arm_ppm_servos[0].set_torque(0)
		else:
			for servo in self._arm_servos:
				servo.set_command(0x0)
			self._arm_ppm_servos[0] = Servo_ppm.Servo_ppm(self._ppm_addr,self.master.bus, 1, self._ppm_start_pos[0])
			self._arm_ppm_servos[1] = Servo_ppm.Servo_ppm(self._ppm_addr,self.master.bus, 2, self._ppm_start_pos[1])

File: cpp/src/Rooky.cpp
#include "Rooky.h"

enum PpmServo
{
	SERVO_1 = 1,
	SERVO_2
};

struct Limit
{
	int min;
	int max;
};

std::vector<int> _servos;

int _ppmServoAddr = 0;
bool _debug = false;

static RookySide _rookySide = RookySide::NONE;

const std::vector<int> LEFT_ARM_SERVOS = { 31, 32, 33, 34, 35 };
const std::vector<int> RIGHT_ARM_SERVOS = { 21, 22, 23, 24, 25 };

std::map<std::string, int> _set_points = { { "joint_2", 0 },
											{ "joint_3", 0 },
											{ "joint_4", 0 },
											{ "joint_5", 0 }, };

const std::map<std::string, float> JointsRatio = { { "joint_1", 3.0 },
													{ "joint_2", 1.9 },
													{ "joint_3", 1.9 },
													{ "joint_4", 1.2 },
													{ "joint_5", 1.2 },
													{ "joint_6", 1.7 },
													{ "joint_7", 0.3 } };

const std::map<std::string, Limit> JointsLimits = { { "joint_1", { -10, 90 } },
													{ "joint_2", { -10, 90 } },
													{ "joint_3", { -90, 90 } },
													{ "joint_4", { -10, 90 } },
													{ "joint_5", { -90, 90 } },
													{ "joint_6", { -20, 35 } },
													{ "joint_7", { 0, 74 } } };

const int _rightArmServoPpm = 26;
const int _leftArmPpmServo = 36;

std::vector<int> _untouchedValues;

bool initRooky(std::string port, RookySide side, bool debug)
{
	std::vector<int> servosAddresses;
	switch (side)
	{
		case LEFT:
			servosAddresses = LEFT_ARM_SERVOS;
			_ppmServoAddr = _leftArmPpmServo;
			break;
		case RIGHT:
			servosAddresses = RIGHT_ARM_SERVOS;
			_ppmServoAddr = _rightArmServoPpm;
			break;
		default:
			_ppmServoAddr = 0;
			ERROR("Initialization failed. Incorrect Rooky side\n")
			;
			return false;
	}

	_rookySide = side;

	if (init_port(port.c_str(), debug) < 0)
	{
		ERROR("Can't init port %s\n", port.c_str());
		return false;
	}

	for (auto addr : servosAddresses)
	{
		initServo(addr);
		//Стандартный ре-им работы сервопривода
		set_PID_mode(addr, NORMAL);

		//Включение питания обмоток двигателя
		set_torque(addr, true);
		_servos.push_back(addr);
	}

	initServoPpm(_ppmServoAddr, 0, side == RookySide::LEFT ? -90 : 90);

	if (!read_touch(_ppmServoAddr, _untouchedValues))
	{
		ERROR("Can't read touch %d\n", _ppmServoAddr);
		return false;
	}
	_debug = debug;
	return true;
}

/**
 * Переводит градусы в тики для сервопривода
 * @param deg градусы для конвертации
 * @return тики
 */
int _degreesToTick(float deg)
{
	return static_cast<int>(deg / 0.0219);
}

/**
 * Проверка позиции на предельные значения
 * @param name - имя сустава
 * @param position - желаемая позиция
 * @return true - позиция находится в пределах
 * 		   false - некоррктная позиция или ошибка
 */
bool _checkLimits(const std::string &name, float position)
{
	auto it = JointsLimits.find(name);

	if (it == JointsLimits.end())
	{
		ERROR("Can't find joint %s\n", name.c_str());
		return false;
	}

	Limit limits = it->second;

	if (position > limits.max || position < limits.min)
	{
		ERROR("Joint \"%s\", invalid setpoint: %f\n", name.c_str(), position);
		return false;
	}
	return true;
}

/**
 * Повернуть прямой сустав, такой один в предплечье
 * такой сустав двигается одним сервоприводом
 * @param name - имя сустава
 * @param speed - предельная скорость движения
 * @param position - желаемая позиция
 * @return
 */
bool _moveSimpleJoint(const std::string &name, float speed, float position)
{
	if (!_checkLimits(name, position))
		return false;

	auto it = JointsRatio.find(name);

	if (it == JointsRatio.end())
	{
		ERROR("Can't load ratio for joint \"%s\"\n", name.c_str());
		return false;
	}

	float ratio = it->second;

	int setpoint = _degreesToTick(position * ratio);
	set_torque(_servos[0], true);
	set_speed_limit(_servos[0], speed);
	set_point(_servos[0], static_cast<int16_t>(setpoint));
	return true;
}

/**
 * Привести в движение дифференциальный сустав (состоит из двух сервоприводов)
 * @param name - имя сустава
 * @param speed - предельная скорость вращения для сервоприводов
 * @return true - все ок
 * 		   false - ошибка
 */
bool _moveDiffJoint(const std::string &name, float speed)
{
	if (name == "joint_2" || name == "joint_3")
	{
		set_speed_limit(_servos[1], speed);
		set_speed_limit(_servos[2], speed);
		set_torque(_servos[1], true);
		set_torque(_servos[2], true);
		set_point(_servos[1], _set_points["joint_2"] + _set_points["joint_3"]);
		set_point(_servos[2], _set_points["joint_2"] - _set_points["joint_3"]);
		return true;
	}

	if (name == "joint_4" || name == "joint_5")
	{
		set_speed_limit(_servos[3], speed);
		set_speed_limit(_servos[4], speed);
		set_torque(_servos[3], true);
		set_torque(_servos[4], true);
		set_point(_servos[3], _set_points["joint_4"] + _set_points["joint_5"]);
		set_point(_servos[4], _set_points["joint_4"] - _set_points["joint_5"]);
		return true;
	}
	DEBUG("exit\n");
	ERROR("Wrong diff joint - %s\n", name.c_str());
	return false;
}

/**
 * Установить дифференциальный сустав в желаемую позицию
 * @param name - имя сустава
 * @param speed - предельная скорость для сервоприводов об/мин
 * @param position - желаемая позиция
 * @return
 */
bool _setDiffJoint(const std::string &name, float speed, float position)
{
	if (!_checkLimits(name, position))
		return false;

	auto it = JointsRatio.find(name);

	if (it == JointsRatio.end())
	{
		ERROR("Can't load ratio for joint \"%s\"\n", name.c_str());
		return false;
	}

	float ratio = it->second;

	_set_points[name] = static_cast<int>(_degreesToTick(position) * ratio);
	_moveDiffJoint(name, speed);
	return true;
}

/**
 * Привести в движение сервопривод второго типа
 * одна управляющая плата с одним адресом на 2 сервопривода
 * @param name - имя сустава
 * @param servoNum - порядковый номер сервопривода на управляющей плате
 * @param position - желаемая позиция в градусах
 * @return
 */
bool _moveServoJiont(const std::string &name, PpmServo servoNum, float position)
{
	DEBUG("Name - %s, position - %f\n", name.c_str(), position);
	if (!_checkLimits(name, position))
		return false;

	auto it = JointsRatio.find(name);

	if (it == JointsRatio.end())
	{
		ERROR("Can't load ratio for joint \"%s\"\n", name.c_str());
		return false;
	}

	float ratio = it->second;

	int angle;
	switch (servoNum)
	{
		case PpmServo::SERVO_1:
			angle = (position * ratio) * -1;
			break;
		case PpmServo::SERVO_2:
			if (_rookySide == RookySide::LEFT)
				angle = -90 + (position * ratio);
			else if (_rookySide == RookySide::RIGHT)
				angle = 90 - (position * ratio);
			else
				return false;
			break;
	}
	DEBUG("Raw servo angle - %d\n", angle);
	set_ppm_angle(_ppmServoAddr, servoNum, static_cast<int16_t>(angle));
	return true;
}

/**
 * Проверка ошибок и их вывод
 * @param addr - адрес, необходим при выводе списка ошибок
 * @param errors - ошибки
 */
void _checkErrors(int addr, std::vector<const char*> errors)
{
	for (auto err : errors)
	{
		if (err != nullptr)
			ERROR("Servo: %d: in troubles: %s \n", addr, err);
	}
}

void moveJoint(const std::string &name, float speed, float position)
{
	if (name == "joint_1")
		_moveSimpleJoint(name, speed, position);
	else if (name == "joint_2")
		_setDiffJoint(name, speed, position);
	else if (name == "joint_3")
		_setDiffJoint(name, speed, position);
	else if (name == "joint_4")
		_setDiffJoint(name, speed, position);
	else if (name == "joint_5")
		_setDiffJoint(name, speed, position);
	else if (name == "joint_6")
		_moveServoJiont(name, PpmServo::SERVO_1, position);
	else if (name == "joint_7")
		_moveServoJiont(name, PpmServo::SERVO_2, position);
	else
		ERROR("Wrong joint name: %s\n", name.c_str());
}

void relax(bool flag)
{
	DEBUG("Relax - %s\n", flag ? "true" : "false");
	if (flag)
	{
		for (auto addr : _servos)
			set_command(addr, 0xAAAA);
		set_torque(_ppmServoAddr, false);
	}
	else
	{
		for (auto addr : _servos)
			set_command(addr, 0x0);
		initServoPpm(_ppmServoAddr, 0, _rookySide == RookySide::LEFT ? -90 : 90);
	}
}

bool readServos(std::vector<ServoData> &servo_datas)
{
	for (auto addr : _servos)
	{
		ServoData data = {};
		if (!get_servo_data(addr, data))
		{
			ERROR("Can't get data from servo %d", addr);
			return false;
		}
		std::vector<const char*> errors;
		for (auto err : data.errors)
			errors.push_back(err);

		_checkErrors(addr, errors);

		servo_datas.push_back(data);
	}
	return true;
}

bool readPpmServo(PpmServoData &data)
{
	PpmServoData dt = {};
	if (!get_ppm_servo_data(_ppmServoAddr, dt))
	{
		ERROR("Can't get data from ppm_servo %d", _ppmServoAddr);
		return false;
	}
	std::vector<const char*> errors;
	for (auto err : dt.errors)
		errors.push_back(err);
	_checkErrors(_ppmServoAddr, errors);

	data = dt;

	return true;
}

bool isTouched()
{
	std::vector<int> vals;
	if (!read_touch(_ppmServoAddr, vals))
		return false;

	DEBUG("vals %d, %d\n", vals[0], vals[1]);
	DEBUG("untouched %d, %d\n", _untouchedValues[0], _untouchedValues[1]);
	if (!vals.size() || !_untouchedValues.size())
		return false;

	if (vals[0] < _untouchedValues[0] * 0.95 || vals[1] < _untouchedValues[1] * 0.95)
		return true;
	return false;
}
